Update connectors doc. (#2308)
feldera/feldera: docs/connectors/index.mdx

## Basics

A connector is a JSON object that describes an external data source or sink such as a database

> table or a Kafka topic. They are specified as a list of JSON connector objects

Reviewer: ```suggestion
table or a Kafka topic. A SQL table can have multiple source connectors attached to it, specified as a list of JSON connector objects
```

under the `'connectors'` attribute in the `WITH` clause of the table.

> Here is an example, where the `VENDOR` has one connector, configured to fetch data in a

Reviewer: ```suggestion
Here is an example, where `VENDOR` has one connector, configured to fetch data in
```

JSON format using HTTP and insert it:

```sql
create table VENDOR (
    id bigint not null primary key,
    name varchar,
    address varchar
) WITH ('connectors' = '[{
    "transport": {
        "name": "url_input", "config": {"path": "https://feldera-basics-tutorial.s3.amazonaws.com/vendor.json"}
    },
    "format": { "name": "json" }
}]');

> ```

Reviewer: Likewise, a SQL view can have one or more sink connectors:

```sql
this needs an example, especially because the syntax for views is slightly different 
```

Author: what's the difference? the with can't be at the end?

Reviewer: yep

A connector specification consists of three parts:

* [Generic attributes](#generic-attributes) common to all connectors, such as backpressure thresholds.
* Transport specification (`transport`), which defines the data transport to be used by the connector
  (e.g.,
  [Kafka](/docs/connectors/sources/kafka), 
  [URL](/docs/connectors/sources/http-get), 

>   [Delta Lake](/docs/connectors/sources/delta), etc.) and its configuration.

Reviewer: See [input](/docs/connectors/sources/) and [output](/docs/connectors/sinks/) connector documentation for a list of supported transports.

## Basics

> A connector is a JSON object that describes an external data source or sink such as a database

Reviewer 2: Users configure connectors using a JSON object, that describes an external.....

table or a Kafka topic. A SQL table can have multiple source connectors attached to it, specified as a list of JSON connector objects
under the `'connectors'` attribute in the `WITH` clause of the table.

Here is an example, where the `VENDOR` table has one connector, configured to fetch and insert JSON data
from an HTTP URL, and a `VENDOR_VIEW` which sends the changes of the view
to a Kafka topic in a format that can be consumed by Debezium:

```sql

> create table VENDOR (

Reviewer 2: We seem to use unusual and inconsistent SQL-styles throughout our examples, docs and posts. Let's stick to the common way of having identifiers in lower case (vendor, not VENDOR), and SQL syntax in upper case. So this would be:

```
CREATE TABLE vendor (
  id BIGINT NOT NULL PRIMARY KEY,
  name VARCHAR,
  some_other_field VARCHAR,
...
) WITH () ...
```

Author: I think the "more common" way is to use uppercase -- but anyways we seem to be using _mostly_ uppercase so far mostly in the docs, we should probably change it all at once during a refactoring

Reviewer 3: this is a religious war that you better not get into
maybe we should build a sqlfmt tool to end the disputes

Reviewer 2: Whatever we do, pick one and stick with it, and preferably, use a formatter. Right now, we use a mix of styles even in the same program and even in the same query!

Reviewer 2: My suggestion for a style guide around casing at least: https://docs.telemetry.mozilla.org/concepts/sql_style

> > **Note:** these basics apply to all connectors **except** the

Reviewer: **Note:** Some transports, e.g., [Delta Lake](/docs/connectors/sources/delta) and [datagen](/docs/connectors/sources/datagen), use fixed predefined data formats and do not require the `format` section in the connector specification.

> > **Note:** these basics apply to all connectors **except** the

Reviewer: See [data format documentation](...) for a list of supported formats.

## Example usage

A program with a view using an output connector that sends changes to a `test_view` Kafka topic in a
JSON format suitable for consumption by Debezium:

```sql
create table test_table(
    id bigint not null primary key,
    f1 boolean,
    f2 string,
    f3 tinyint,
    f4 decimal(5,2),
    f5 float64,
    f6 time,
    f7 timestamp,
    f8 date,
    f9 binary
);

create view test_view
WITH (
    'connectors' = '[{
        "max_queued_records": 1000,
        "format": {
            "name": "json",
            "config": {
                "update_format": "debezium"
            }
        },
        "transport": {
            "name": "kafka_output",
            "config": {
                "bootstrap.servers": "redpanda:9092",
                "topic": "test_view"
            }
        }
    }]'
)
as select * from test_table;
```

## Additional resources

For more information, see:

> * [Tutorial using HTTP-based Input and Output](/docs/tutorials/basics/part2)

Reviewer: ```suggestion
* [Tutorial on using input and output connectors](/docs/tutorials/basics/part3)
* [Tutorial on using HTTP-based Input and Output](/docs/tutorials/basics/part2)
```

* [Supported transport for Tables/Sources](/docs/connectors/sources)

> * [Supported transport for Views/Sinks](/docs/connectors/sinks)

Reviewer: ```suggestion
* [Supported sink transports](/docs/connectors/sinks)
```

> * [Supported transport for Views/Sinks](/docs/connectors/sinks)

Reviewer: * [Supported data formats]()
// There's currently no top-level page for formats that can be referenced here. Can we add it?
